californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java

```diff
@@ -118,28 +118,47 @@ public void proactiveCancel() {
 		// copy options, but set Observe to cancel
 		cancel.setOptions(request.getOptions());
 		cancel.setObserveCancel();
-		
+
 		// dispatch final response to the same message observers
-		for (MessageObserver mo: request.getMessageObservers()) {
+		for (MessageObserver mo : request.getMessageObservers()) {
 			cancel.addMessageObserver(mo);
 		}
-		
+
 		endpoint.sendRequest(cancel);
-		
-		// cancel old ongoing request
+	}
+
+	/** 
+	 * Cancel observer. 
+	 */
+	private void cancel() {
 		request.cancel();
 		endpoint.cancelObservation(request.getToken());
 		setCanceled(true);
 	}
 	
 	/**
+	 * Proactive Observe cancellation:
+	 * Cancel the observe relation by sending a GET with Observe=1.
+	 */
+	public void proactiveCancel() {
+		sendCancelObserve();
+
+		// cancel old ongoing request
+		cancel();
+	}
+	
+	/**
 	 * Reactive Observe cancellation:
 	 * Cancel the observe relation by forgetting, which will trigger a RST.
+	 * For TCP, {{@link #proactiveCancel()} will be executed.
 	 */
 	public void reactiveCancel() {
-		request.cancel();
-		endpoint.cancelObservation(request.getToken());
-		setCanceled(true);
+		if (CoAP.isTcpScheme(request.getScheme())) {
```

Comment:
Assuming a client doesn't know which protocol is in use, then I guess it has a certain expection of what should happen when it invokes this method, right? Otherwise it would/should have invoked `proactiveCancel()`, shouldn't it?